ENT-3562: New StringToLong functions

olehermanse opened the pull request:
Must be merged together:
https://github.com/cfengine/core/pull/2990
https://github.com/cfengine/enterprise/pull/359
https://github.com/cfengine/nova/pull/1073

olehermanse commented:
Jenkins:
https://ci.cfengine.com/view/all/job/bootstrap-enterprise-pr/549/

olehermanse commented:
@jimis @craigcomstock Can you please review these 3 pull requests. Might include some subtle change in behavior, good to get a second (and third) opinion.

jimis commented on libutils/string_lib.c:
```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
+ * @param[out] number Where to store result on success, cannot be NULL
+ * @return 0 on success, error code otherwise (see source code)
+ */
+int StringToLong(const char *str, long *number)
```
I would suggest to rename `number` to `value_out` in both the C and header file so that it's obvious this is an "out" parameter.

olehermanse replied:
Thanks, done.

jimis commented on libutils/string_lib.c:
```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
+/**
+ * @brief Converts a string of numerals in base 10 to a long integer.
+ *
+ * Result is stored in *number, return value should be chacked.
```
`chacked` typo

olehermanse replied:
Thanks, fixed

jimis commented on libutils/string_lib.c:
```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
+ * @brief Converts a string of numerals in base 10 to a long integer.
+ *
+ * Result is stored in *number, return value should be chacked.
+ * On error *number is unmodified and -1 is returned.
```
Not -1, but "an error code is returned".

I'm not sure if this is the best way though. You might as well return bool type and `false` on any error. Still I'm split on this, so error code is OK.

olehermanse replied:
I like int better because it's easier to debug, also makes it easier to check for integer overflow for example.

jimis commented on libutils/string_lib.h:
```diff
@@ -59,7 +59,12 @@ char ToUpper(char ch);
 
-long StringToLong(const char *str);
+int StringToLong(const char *str, long *number) FUNC_WARN_UNUSED_RESULT;
+void LogConversionError(const char *str_attempted, const char *id);
```
Why export this function. Do you use it in nova? If not, why even have a function?

olehermanse replied:
Yes it is used in nova. To avoid code duplication.

jimis commented on libutils/string_lib.c:
```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
+ * @param[in] str String with numerals to convert, cannot be NULL
+ * @return Result of conversion
+ */
+long StringToLongUnsafe(const char *str)
```
I thought you replaced this everywhere? Or where there too many occurences after all?

olehermanse replied:
If you think all of the places where this function has been replaced with new functions (with slightly differing behavior) are good, we can remove the old version.

jimis replied:
Since you are triggering builds like crazy, why don't you add a commit that removes it and trigger another, to see if/what breaks? :-)

olehermanse replied:
Nothing will break if I remove it, it is unused.

I am asking you to review the places where I've made changes by replacing the old StringToLong with new, different, functions.

jimis replied:
There are very few places in this pull request and I checked them, they look OK.

olehermanse commented:
@jimis I have added your suggestions and started a new jenkins build. (Link in previous comment updated). As you know, I'm exhausting the jenkins queue daily.

vpodzime commented on libutils/string_lib.c:
```diff
@@ -372,13 +372,140 @@ bool EmptyString(const char *s)
+ */
+void LogConversionError(const char *str_attempted, const char *id)
+{
+    Log(LOG_LEVEL_ERR, "Conversion error on '%s' (%s)", str_attempted, id);
```
Would be nice if this also took the return code and printed the respective error message based on it.

olehermanse replied:
That would make it specific to StringToLong, but I agree. I'll add this.

olehermanse commented:
@jimis @vpodzime @craigcomstock I believe I have addressed all your comments. I'll start one more build in jenkins now, believing this is ready to merge.

https://ci.cfengine.com/view/all/job/bootstrap-enterprise-pr/552/

olehermanse commented:
Green build - merging.
